state/execution.go

```diff
@@ -373,7 +391,7 @@ func updateState(
 	// Update the validator set with the latest abciResponses.
 	lastHeightValsChanged := state.LastHeightValidatorsChanged
 	if len(abciResponses.EndBlock.ValidatorUpdates) > 0 {
-		err := updateValidators(nValSet, abciResponses.EndBlock.ValidatorUpdates)
+		err := updateValidators(nValSet, abciResponses.EndBlock.ValidatorUpdates, state.ConsensusParams.Validator)
```

Comment:
can we not mix validation and update logic? maybe extract validation into a separate function

```go
if err := validateUpdates(abciResponses.EndBlock.ValidatorUpdates, state.ConsensusParams.Validator.ValidatorPubkeyTypes); err != nil {
    return state, fmt.Errorf("Error in validator updates: %v", err)
}
```

Reply:
Great point! I was originally trying to avoid an extra iteration over the validator set, but I think separating concerns is better for code clarity.